Refactor middleware

dhvcc opened the pull request:
closes #328 

## Прогресс

- [X] Рефактор
- [X] Типы
- [X] Тесты
- [X] Документация

## Изменения

### Новый синтаксис мидлварей

1. Остановка обработки / обновление контекста
```python3
async def pre(self) -> None:
    if some_error:
        self.stop("Some error occurred")
    else:
         self.send({"new_context_key": "new value"})
```
2. Запуск `pre` и `post` методов
```python3
class SomeEventView(ABCView):
    ...

    async def handle_event(...):
        mw_instance = middleware(event) # Создаем инстанс
        await mw_instance.pre() # Запускаем pre

        if mw_instance.can_forward: # Были ли ошибки при выполнени?
            some_logger.info(f"New context received {new_context}")
            some_context.update(new_context)
        elif new_context := mw_instance.context_update: # Нужно ли обновить контекст?
            some_logger.warning(f"Middleware stopped handling, exception args {mw_instance.error.args}")
            some.logger.exception(mw_instance.error)
            return

        ...

        await mw_instance.post(...)
```
3. Новый класс исключений - `MiddlewareError` 


### Новый сахар для `ABCView`

```python3
class SomeEventView(ABCView):
    ...

    async def handle_event(
        self, event: dict, ctx_api: "ABCAPI", state_dispenser: "ABCStateDispenser"
    ) -> Any:
        # Handle setup
        context_variables: dict = {}
        ...
        # Handle setup

        if await self.pre_middleware(event, context_variables):
            return # pre_middleware возвращает exception или None

        # Run handlers        
        handler_responses = []
        handlers = []
        ...
        # Run handlers

        await self.post_middleware(self, handler_responses, handlers)
```

dhvcc committed:
Refactored Middleware

Updated middleware class, now need to register the class, not the
instance

Added pre_middleware and post_middleware methods to ABCView

dhvcc committed:
Fixed mypy issues, imports, MiddlewareError class

Renamed "error" wrapper method to "stop", to remove confusion

timoniq reviewed (approved):
Спасибо, вроде бы все в порядке, но есть некоторые комментарии:

Раньше мидлварям можно было делать кастомный инит, передавать какие-то параметры прямо при регистрации во view - может стоит придумать что-нибудь

В некоторых местах тайпхинты странные (`var: dict = {}` (зачем это?), `request_text(...) -> Union[str, Any]`)

dhvcc committed:
Added middleware tests/fixtures, fixed warnings

Reraising if .stop argument is already an exception

dhvcc committed:
Updated tests, middleware list is now Set

Added super calls

dhvcc committed:
Updated Makefile, fixed with make fix

Using "is" to assert for None/False in tests
Ignoring E203 as it's not PEP8 compliant

dhvcc committed:
Updated docs and pre_middleware view method
